modules/rocksdb-common/src/main/java/org/apache/ignite/internal/rocksdb/snapshot/RocksSnapshotManager.java

```diff
@@ -0,0 +1,212 @@
+import java.util.concurrent.Executor;
+import org.apache.ignite.internal.rocksdb.RocksUtils;
+import org.apache.ignite.internal.util.IgniteUtils;
+import org.apache.ignite.lang.IgniteInternalException;
+import org.rocksdb.EnvOptions;
+import org.rocksdb.IngestExternalFileOptions;
+import org.rocksdb.Options;
+import org.rocksdb.ReadOptions;
+import org.rocksdb.RocksDB;
+import org.rocksdb.RocksDBException;
+import org.rocksdb.RocksIterator;
+import org.rocksdb.Slice;
+import org.rocksdb.Snapshot;
+import org.rocksdb.SstFileWriter;
+
+/**
+ * Class for creating and restoring RocksDB snapshots.
+ */
+public class RocksSnapshotManager {
+    /** Suffix for the temporary snapshot folder. */
+    private static final String TMP_SUFFIX = ".tmp";
+
+    private final RocksDB db;
+
+    private final Collection<ColumnFamilyRange> ranges;
+
+    private final Executor executor;
+
+    /**
+     * Creates a new instance of the snapshot manager.
+     * This instance <b>does not</b> own any of the provided resources and will not close them.
+     *
+     * @param db RocksDB instance which snapshots will be managed.
+     * @param ranges Key ranges of Column Families that exist in the provided {@code db} instance.
+     * @param executor Executor which will be used for creating snapshots.
+     */
+    public RocksSnapshotManager(RocksDB db, Collection<ColumnFamilyRange> ranges, Executor executor) {
+        assert !ranges.isEmpty();
+
+        this.db = db;
+        this.ranges = ranges;
+        this.executor = executor;
+    }
+
+    /**
+     * Creates a snapshot of the enclosed RocksDB instance and saves it into a provided folder.
+     *
+     * @param snapshotDir Folder to save the snapshot into.
+     * @return Future that either completes successfully upon snapshot creation or signals a failure.
+     */
+    public CompletableFuture<Void> createSnapshot(Path snapshotDir) {
+        Path tmpPath = Paths.get(snapshotDir.toString() + TMP_SUFFIX);
+
+        return CompletableFuture.supplyAsync(db::getSnapshot, executor)
+                .thenComposeAsync(snapshot -> {
+                    createTmpSnapshotDir(tmpPath);
+
+                    // Create futures for capturing SST snapshots of the column families
+                    CompletableFuture<?>[] sstFutures = ranges.stream()
+                            .map(cf -> createSstFileAsync(cf, snapshot, tmpPath))
+                            .toArray(CompletableFuture[]::new);
+
+                    return CompletableFuture.allOf(sstFutures).thenApply(v -> snapshot);
+                }, executor)
+                .whenCompleteAsync((snapshot, e) -> {
+                    // Release the snapshot
+                    db.releaseSnapshot(snapshot);
+
+                    // Snapshot is not actually closed here, because a Snapshot instance doesn't own a pointer, the
+                    // database does. Calling close to maintain the AutoCloseable semantics
+                    snapshot.close();
```

Comment:
It seems that if `e` is not null, then `snapshot` is null, so this might throw an NPE.

Reply:
nice catch!